### PR TITLE
refactor: add iterator for traversing an AsyncAPI doc

### PR DESCRIPTION
**Pull request by smoya**
**Description**
This PR adds an iterator for traversing AsyncAPI documents. It is based in the current one. See https://github.com/asyncapi/parser-js/blob/master/lib/iterators.js

**Related issue(s)**
https://github.com/asyncapi/parser-js/issues/481

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -58,7 +58,6 @@ rules:
   no-empty-character-class: 2
   no-self-compare: 2
   valid-typeof: 2
-  no-unused-vars: [2, { "args": "none" }]
```
Comment (smoya) [inline .eslintrc:61]: Not compatible with Typescript, however we also use the compatible version so I'm just removing the support here.

File: src/iterator.ts [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    Definitions = 'definitions', // Crawl all schemas in definitions field
+}
+   
+export type TraverseOptions = {
+   callback: Function
+   schemaTypesToIterate: SchemaTypesToIterate[]
+   seenSchemas: Set<any>
+}   
+  
+export type TraverseCallback = (schema: SchemaInterface, propOrIndex: String | Number | null, callbackType: SchemaIteratorCallbackType) => void
```
Comment (magicmatatjahu) [inline src/iterator.ts:46]:
```suggestion
export type TraverseCallback = (schema: SchemaInterface, propOrIndex: string | number | null, callbackType: SchemaIteratorCallbackType) => void
```

File: src/iterator.ts [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+      traverseMessageTrait(messageTrait, options);
+    });
+  }
+}
+
+/* eslint-disable sonarjs/cognitive-complexity */
+/**
+ * Traverse current schema and all nested schemas.
+ */
+function traverseSchema(schema: SchemaInterface, propOrIndex: String | Number | null, options: TraverseOptions) { // NOSONAR
```
Comment (magicmatatjahu) [inline src/iterator.ts:85]:
```suggestion
function traverseSchema(schema: SchemaInterface, propOrIndex: string | number | null, options: TraverseOptions) { // NOSONAR
```

**Review by magicmatatjahu — COMMENTED** [review]
Only two comments :)

**Comment by smoya** [comment]
@magicmatatjahu would you mind revieweing again? I fixed the linting issues but also improved the tests. Thanks!

**Review by magicmatatjahu — APPROVED** [review]

**Pull request merged**